scripts/release_notes/commitlist.py

```diff
@@ -0,0 +1,278 @@
+
+    @staticmethod
+    def from_existing(path):
+        commits = CommitList.read_from_disk(path)
+        return CommitList(path, commits)
+
+    @staticmethod
+    def create_new(path, base_version, new_version):
+        if os.path.exists(path):
+            raise ValueError("Attempted to create a new commitlist but one exists already!")
+        commits = CommitList.get_commits_between(base_version, new_version)
+        return CommitList(path, commits)
+
+    @staticmethod
+    def read_from_disk(path):
+        with open(path) as csvfile:
+            reader = csv.reader(csvfile)
+            rows = list(reader)
+        assert all(len(row) >= 4 for row in rows)
+        return [Commit(*row[:4]) for row in rows]
+
+    def write_to_disk(self):
+        path = self.path
+        rows = self.commits
+        directory = os.path.dirname(path)
+        os.makedirs(directory, exist_ok=True)
+        with open(path, "w") as csvfile:
+            writer = csv.writer(csvfile)
+            for commit in rows:
+                writer.writerow([commit.commit_hash, commit.category, commit.topic, commit.title])
+
+    def keywordInFile(file, keywords):
+        for key in keywords:
+            if key in file:
+                return True
+        return False
+
+    @staticmethod
+    def categorize(commit_hash, title):
+        features = get_features(commit_hash, return_dict=True)
+        title = features["title"]
+        labels = features["labels"]
+        category = "Uncategorized"
+        topic = "Untopiced"
+
+        # We ask contributors to label their PR's appropriately
+        # when they're first landed.
+        # Check if the labels are there first.
+        already_categorized = already_topiced = False
+        for label in labels:
+            if label.startswith("release notes: "):
+                category = label.split("release notes: ", 1)[1]
+                already_categorized = True
+            if label.startswith("topic: "):
+                topic = label.split("topic: ", 1)[1]
+                already_topiced = True
+        if already_categorized and already_topiced:
+            return Commit(commit_hash, category, topic, title)
+
+        if "deprecation" in title.lower():
+            topic = "deprecations"
+
+        files_changed = features["files_changed"]
+        for file in files_changed:
+            if CommitList.keywordInFile(file, ["docker/", ".github", "packaging/"]):
+                category = "releng"
+                break
+            if CommitList.keywordInFile(
+                file,
+                [
+                    "torchdata/dataloader2",
+                ],
+            ):
+                category = "dataloader2"
+                break
+            if CommitList.keywordInFile(
+                file,
+                [
+                    "torchdata/datapipes",
+                ],
+            ):
+                category = "datapipe"
+                break
```

Comment:
Here are the main changes to adapt the structure of TorchData.
Please let me know if you have any recommendation on the categorization.

Reply:
I think we can also consider mandatory labelling of every PR explicitly (like PyTorch Core does). It will be more accurate that, though take may be 30 sec extra per PR

Reply:
Yeah. That's the reason I add a few labels like PyTorch Core. And, this script does respect such topic label.
For mandatory labelling, I don't have time to do it right now. Treat it as a follow-up task.